omegaconf/base.py

```diff
@@ -325,117 +311,194 @@ def _select_impl(
-            parent, last_key, value = root_node._select_impl(
-                inter_key,
-                throw_on_missing=throw_on_missing,
-                throw_on_resolution_failure=throw_on_resolution_failure,
-            )
-
-            # if parent is None or (value is None and last_key not in parent):  # type: ignore
-            if parent is None or value is None:
-                if throw_on_resolution_failure:
-                    raise ConfigKeyError(
-                        f"{inter_type} interpolation key '{inter_key}' not found"
-                    )
-                else:
-                    return None
-            assert isinstance(value, Node)
-            return value
+        resolver = OmegaConf.get_resolver(inter_type)
+        if resolver is not None:
+            root_node = self._get_root()
+            try:
+                value = resolver(root_node, inter_args, inter_args_str)
+                return ValueNode(
+                    value=value,
+                    parent=self,
+                    metadata=Metadata(
+                        ref_type=None, object_type=None, key=key, optional=True
+                    ),
+                )
+            except Exception as e:
+                self._format_and_raise(key=inter_args, value=None, cause=e)
+                assert False
         else:
-            resolver = OmegaConf.get_resolver(inter_type)
-            if resolver is not None:
-                root_node = self._get_root()
-                try:
-                    value = resolver(root_node, inter_key)
-                    return ValueNode(
-                        value=value,
-                        parent=self,
-                        metadata=Metadata(
-                            ref_type=None, object_type=None, key=key, optional=True
-                        ),
-                    )
-                except Exception as e:
-                    self._format_and_raise(key=inter_key, value=None, cause=e)
-                    assert False
+            if throw_on_resolution_failure:
+                raise UnsupportedInterpolationType(
+                    f"Unsupported interpolation type {inter_type}"
+                )
             else:
-                if throw_on_resolution_failure:
-                    raise UnsupportedInterpolationType(
-                        f"Unsupported interpolation type {inter_type}"
-                    )
-                else:
-                    return None
+                return None
 
     def _resolve_interpolation(
         self,
+        parent: Optional["Container"],
         key: Any,
         value: "Node",
         throw_on_missing: bool,
         throw_on_resolution_failure: bool,
     ) -> Any:
+        value_kind = get_value_kind(value)
+        if value_kind != ValueKind.INTERPOLATION:
+            return value
+
+        parse_tree = parse(_get_value(value))
+        return self._resolve_complex_interpolation(
```

Comment:
looks like you are always calling _resolve_complex_interpolation to resolve an interpolation, however the doc of _resolve_complex_interpolation says that it should only be used for complex interpolations. can you clarify?

Reply:
You're right, this was confusing (I think it's legacy from an older version, can't remember for sure)

Clearer in 2ad3dca?